internal/cuckoo/cuckoo.go

```diff
@@ -0,0 +1,239 @@
+package cuckoo
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/optable/match/internal/hash"
+)
+
+const (
+	// Nhash is the number of hash function used for cuckoo hash
+	Nhash = 3
+	// Maximum number of reinsertions.
+	// Each reinsertion kicks off 1 egg (item) and replace it
+	// with the item being reinserted, and then reinserts the
+	// kicked off egg
+	ReInsertLimit = 200
+	Factor        = 1.4
+	HashFunc      = hash.Highway
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+// Cuckoo represents a 3-way Cuckoo hash table data structure
+// that contains the items, bucket indices of each item and the 3
+// hash functions. The bucket lookup is a lookup table on items which
+// tells us which item should be in the bucket at that index. Upon
+// construction the items slice has an additional nil value prepended
+// so the index of the Cuckoo.items slice is +1 compared to the index
+// of the input slice you use.
+type Cuckoo struct {
+	items        [][]byte
+	hashIndices  []byte
+	bucketLookup []uint64
+	// Total bucket count, len(bucket)
+	bucketSize uint64
+	// 3 hash functions h_0, h_1, h_2
+	hashers [Nhash]hash.Hasher
+}
+
+// NewCuckoo instantiates a Cuckoo struct with a bucket of size Factor * size,
+// a stash and 3 seeded hash functions for the 3-way cuckoo hashing.
+func NewCuckoo(size uint64, seeds [Nhash][]byte) *Cuckoo {
+	bSize := max(1, uint64(Factor*float64(size)))
+	var hashers [Nhash]hash.Hasher
+	for i, s := range seeds {
+		hashers[i], _ = hash.New(HashFunc, s)
+	}
+
+	return &Cuckoo{
+		// extra element is "keeper" to which the bucketLookup can be directed
+		// when there is no element present in the bucket.
+		items:        make([][]byte, size+1),
+		hashIndices:  make([]byte, size+1),
+		bucketLookup: make([]uint64, bSize),
+		bucketSize:   bSize,
+		hashers:      hashers,
+	}
+}
+
+// NewDummyCuckoo instantiates a Cuckoo struct that does not allocate buckets.
+func NewDummyCuckoo(size uint64, seeds [Nhash][]byte) *Cuckoo {
```

Comment:
I would replace this with a CuckooHasher that contains 3 hashers and a bucketSize and returns the bucketIndices. CuckooHash would contain a CuckooHasher. You can embed the cuckoohasher in the cuckoo and you will get automatic delegation.